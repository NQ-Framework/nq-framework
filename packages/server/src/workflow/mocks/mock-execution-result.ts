import { WorkflowExecutionResult } from '@nqframework/models';

export const mockExecutionResult: WorkflowExecutionResult = {
    data: {
        mockProp: "mock value"
    },
    finalOutput: {
        mockFinalOutputProp: "mock final value"
    }
}