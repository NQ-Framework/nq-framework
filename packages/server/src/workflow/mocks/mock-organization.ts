import { Organization } from '@nqframework/models';

export const mockOrganization: Organization = {
    name: "mock org id",
    members: [{
        displayName: "mock display name",
        email: "mock@email.com",
        role: "admin",
        status: "accepted",
        uid: "mock user id"
    }],
    memberIds: ["mock user id"],
    address1: "addr1",
    address2: "addr2",
    country: "ctry",
    dataCredentials: [],
    email: "mock-org@email.com",
    serviceAccounts: [],
    dataSources: []
}