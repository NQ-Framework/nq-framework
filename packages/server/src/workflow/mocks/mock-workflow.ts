import { Workflow, WorkflowTriggerInstance } from '@nqframework/models';

export const mockWorkflow: Workflow = {
  id: 'mock id',
  isActive: true,
  name: 'mock workflow',
  organizationId: 'mock org id',
  actionInstances: [
    {
      name: 'mock action instance id',
      isEnabled: true,
      editorConfig: {
        color: '#000000',
        x: 100,
        y: 100,
      },
      configuration: {
        input: [],
      },
      action: {
        id: 'mock action id',
        isEnabled: true,
        name: 'mock action',
        path: 'mock-path',
        version: 1,
        hasDefaultPort: true,
      },
    },
    {
      name: 'mock action instance id 2',
      isEnabled: true,
      editorConfig: {
        color: '#000000',
        x: 100,
        y: 100,
      },
      configuration: {
        input: [],
      },
      action: {
        id: 'mock action id',
        isEnabled: true,
        name: 'mock action',
        path: 'mock-path',
        version: 1,
        hasDefaultPort: true,
      },
    },
  ],
  actionLinks: [
    {
      fromName: 'mock action instance id',
      toName: 'mock action instance id 2',
      isEnabled: true,
    },
  ],
  triggers: [
    {
      actions: ['mock action instance id'],
      editorConfig: {
        color: '#000000',
        x: 100,
        y: 100,
      },
      id: '123',
      input: [
        { name: 'verb', value: ['GET'] },
        { name: 'endpoint', value: 'test endpoint' },
      ],
      output: [],
      type: 'api',
    } as WorkflowTriggerInstance,
  ],
};
