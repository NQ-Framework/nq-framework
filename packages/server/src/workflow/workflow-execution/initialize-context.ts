import { WorkflowExecutionContext } from '@nqframework/models';

export function initializeContext(
  context: WorkflowExecutionContext,
): WorkflowExecutionContext {
  const workflow = context.workflow;
  const actionInstances = workflow.actionInstances
    ? workflow.actionInstances.filter((ai) => ai.isEnabled)
    : [];

  if (actionInstances.length === 0) {
    throw new Error(
      'Execution context cannot be initialized with zero actions',
    );
  }

  let actionLinks = workflow.actionLinks;
  if (!actionLinks) {
    actionLinks = [];
  } else {
    actionLinks = actionLinks.filter((al) =>
      al.isEnabled &&
      actionInstances.some((ai) => ai.name === al.fromName) &&
      actionInstances.some((ai) => ai.name === al.toName)
    );
  }
  return {
    ...context,
    workflow: { ...workflow, actionInstances, actionLinks },
  };
}
