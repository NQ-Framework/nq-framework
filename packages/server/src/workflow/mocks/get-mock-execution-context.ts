import { WorkflowExecutionContext } from '@nqframework/models';
import { mockOrganization } from './mock-organization';
import { mockWorkflow } from './mock-workflow';

export const getMockExecutionContext = (): WorkflowExecutionContext =>
  JSON.parse(
    JSON.stringify({
      actions: {
        'mock name': {
          properties: [{ name: 'mock prop', value: 'mock prop val' }],
          values: [{ name: 'mock value ', value: 'mock value val' }],
        },
      },
      input: { mockInput: 'mock input value' },
      id: 'mock execution context id',
      startedBy: 'mock user id',
      organization: mockOrganization,
      isRunning: true,
      stack: [],
      startTime: new Date('2020-11-20'),
      triggerInput: [{ name: 'trigInput', value: 'mock trigger input value' }],
      triggerOutput: { output: 'mock trigger output' },
      workflow: mockWorkflow,
    } as WorkflowExecutionContext),
  ) as WorkflowExecutionContext;
