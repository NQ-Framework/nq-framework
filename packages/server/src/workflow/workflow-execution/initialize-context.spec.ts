import { getMockExecutionContext } from '../mocks/get-mock-execution-context';
import { initializeContext } from "./initialize-context";

describe("initializeContext", () => {
    it('sets links to empty array if undefined', () => {
        const result = initializeContext({ ...getMockExecutionContext(), workflow: { ...getMockExecutionContext().workflow, actionLinks: undefined as any } });
        expect(result.workflow.actionLinks).toBeDefined();
        expect(result.workflow.actionLinks.length).toBe(0);
    });

    it('throw if workflow has no actions', () => {
        expect(() => {
            initializeContext({ ...getMockExecutionContext(), workflow: { ...getMockExecutionContext().workflow, actionInstances: undefined as any } });
        }
        ).toThrowErrorMatchingSnapshot();
    });

    it('removes disabled action links', () => {
        const result = initializeContext({
            ...getMockExecutionContext(),
            workflow: {
                ...getMockExecutionContext().workflow,
                actionLinks: [
                    {
                        ...getMockExecutionContext().workflow.actionLinks[0],
                        isEnabled: false
                    }
                ]
            }
        });
        expect(result.workflow.actionLinks.length).toBe(0);
    });
    it('removes invalid action links', () => {
        const result = initializeContext({
            ...getMockExecutionContext(),
            workflow: {
                ...getMockExecutionContext().workflow,
                actionLinks: [
                    {
                        ...getMockExecutionContext().workflow.actionLinks[0],
                        fromName: 'incorrect name'
                    },
                    {
                        ...getMockExecutionContext().workflow.actionLinks[0],
                        toName: 'incorrect name'
                    },
                    getMockExecutionContext().workflow.actionLinks[0],
                ]
            }
        });
        expect(result.workflow.actionLinks.length).toBe(1);
    });
    it('removes disabled action instances', () => {
        const result = initializeContext({
            ...getMockExecutionContext(),
            workflow: {
                ...getMockExecutionContext().workflow,
                actionInstances: [
                    {
                        ...getMockExecutionContext().workflow.actionInstances[0],
                        isEnabled: false
                    },
                    getMockExecutionContext().workflow.actionInstances[1]
                ]
            }
        });
        expect(result.workflow.actionInstances.length).toBe(1);
    });
})