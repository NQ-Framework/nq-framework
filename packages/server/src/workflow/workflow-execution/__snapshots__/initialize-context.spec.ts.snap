// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializeContext throw if workflow has no actions 1`] = `"Execution context cannot be initialized with zero actions"`;
